src/qibolab/sequence.py

```diff
@@ -93,17 +94,36 @@ def pulse_channels(self, pulse_id: int) -> list[ChannelId]:
         """Find channels on which a pulse with a given id plays."""
         return [channel for channel, pulse in self if pulse.id == pulse_id]
 
-    def concatenate(self, other: "PulseSequence") -> None:
+    def concatenate(self, other: Iterable[_Element]) -> None:
         """Juxtapose two sequences.
 
         Appends ``other`` in-place such that the result is:
             - ``self``
             - necessary delays to synchronize channels
             - ``other``
         """
-        _synchronize(self, other.channels)
+        _synchronize(self, PulseSequence(other).channels)
```

Comment:
I think here you need
```suggestion
        _synchronize(self, PulseSequence(other).channels | self.channels)
```
otherwise the delay is not added.

Here is a simple test to verify:
```py
platform = create_platform("dummy")
q0 = platform.natives.single_qubit[0]

sequence = PulseSequence()
sequence |= q0.RX()
sequence |= q0.MZ()

assert isinstance(sequence[0][1], Pulse)
assert isinstance(sequence[1][1], Delay)
assert isinstance(sequence[2][1], Readout)
```

Reply:
Here https://github.com/qiboteam/qibolab/issues/1003#issuecomment-2320387728.

Reply:
Correct, and I forgot checking it again.

Sorry, my bad 🙇

Reply:
@stavros11 I better investigated, and it was false, it is correct without the `self.channels`.

My I didn't correctly express during the meeting, but the role of `concatenate` is to make to concatenated sequence starting altogether, but also as soon as possible.
I.e. if a sequence contains multiple pulses starting at the same time, at the beginning of the sequence, they should start at the same time even in the concatenated sequence.
On the other hand, if two sequences playing on non-overlapping channels are concatenated, they should start at the same time.

This second behavior could be desired in many cases, since pulses should be able to start in parallel. It is definitely desired in case of pulses (including gates) played on different qubits (or fully separate sets of qubits), but it is not for gates on the same qubit, which should be played in sequence (as in your example https://github.com/qiboteam/qibolab/issues/1003#issuecomment-2320387728).
However, the behavior for gates on the same qubit involving completely different channels (like `RX` and `MZ`) is enforced by the compiler (as it requires a more advanced concept than the channel, i.e. the `Qubit`, which is not sufficiently simple to be included at the level of the `PulseSequence` and plain natives, requiring instead a `Circuit`).

However, I see the confusion which is generated, and `|` is not even the best simple, as it suggests juxtaposition (which is now achieved including `| self.channels`).
My proposal is then to keep both: let's keep `concatenate`, but with a different operator alias (I'd take `<<`, inspired to C++ streams), and another method, say `juxtapose`, which is the one implemented in the suggestion, taking instead the `|` alias.

The last bit would be to add a bit of syntactic sugar for the compiler as well, in order to make it simpler to interleave sequences and circuits.
E.g. we could support `PulseSequence` additions to a wrapped circuit, which in turn will compile the circuit to a pulse sequence, and then perform juxtaposition.
A possible example syntax could be:
```py
seq = (NativeCircuit(5, compiler=my_compiler) 
        | q0.RX() 
        | q3.RX12()
        << [("q0/drive", Delay(5))]
        | p03.CNOT()
        << [("q0/drive", Align()), ("q2/drive", Align())]
        + [X(2), X(3), Z(1)]
        | p23.CNOT()
        + [Z(q) for q in range(5)]
        + M(*range(5))
    )
```
where a wrapped circuit is initialized with a number of qubits and equipped with a compiler, but in practice maintaining internally a `PulseSequence`, just supporting `Circuit` additions through compilation (or better, `Union[Circuit, Iterable[Gate]]`, since we just need a queue).
We could even expand this objects with further advanced sugar (e.g. `@ ("q0/drive", "q2/drive")` to align channels, or `@ (0, 2, 3)` to directly align qubits), since the object will have everything at its disposal.

In any case, I'd fix the low-level interface in this PR (both `concatenate` and `juxtapose`, in the hope it will reduce the confusion), including their aliases (since they are just tiny shortcuts), and postpone anything more elaborate to future releases.

Reply:
I was also not sure what `concatenate` was supposed to achieve, that's why I didn't comment about it in the PR that first implemented it. Then in the meeting when we presented the sequence building shortcuts, you said that in qibocal we should be able to do things like
```py
sequence = PulseSequence()
sequence |= q0.RX()
sequence |= q0.MZ()
```
so I assumed that this is expected to add the delays, which does not happen unless you use all channels.

> My proposal is then to keep both: let's keep `concatenate`, but with a different operator alias (I'd take `<<`, inspired to C++ streams), and another method, say `juxtapose`, which is the one implemented in the suggestion, taking instead the `|` alias.

I am fine with this proposal, however do you have in mind any practical application where `concatenate` (`<<`) would actually be useful? I expect that `juxtapose` (`|`) will be useful in most qibocal routines, since we are usually doing something in the drive/pulse and then we want to measure after. I cannot think of something similar for `concatenate`, given that we already have the default `append` and `extend` which already let the user do whatever they want, including playing pulses in parallel (on different channels). In principle, even `append` on its own should be sufficient, all the rest are just helpers.

Reply:
I expect `concatenate` to be useful when you play with multiple qubit at once, so you want to operate them in parallel. E.g. the ZZ protocol @andrea-pasquale started writing is not an exact match, since it is juxtaposed, but in principle you could even decide to play it in parallel, or adding your delay manually (and possibly even sweeping it).

Of course, the basic list methods are sufficient, since `PulseSequence` is nothing else than a `list` subclass.
The added value of `juxtapose` is clear, since it frequently happens to play natives one after the other, and you definitely want to pad them. Instead, `concatenate` is less clear, since imposing to start as soon as possible but altogether is almost a short-cut for a conditional (if you assume that the channels of the two sequences either overlap or not, it's deciding whether to start the two concatenated sequences in parallel or juxtapose them - though they could even partially and the duration of the overlapping channels in the first sequence is not the full duration of the first sequence, and the second will start half the way).

In any case, you should interpret them about guarantees given:
- `extend` gives no guarantees, or better, it guarantees that no `Delay` is added
- `concatenate` only guarantees that the all the channels in the `concatenated` sequence will start simultaneously
- `juxtapose` will guarantee simultaneous start and no overlap
- the compiler is similar to `juxtapose` (`initial` padding), but moments-aware and with extra guarantees on qubits involved in interactions (`final` padding)

Reply:
> My proposal is then to keep both: let's keep `concatenate`, but with a different operator alias (I'd take `<<`, inspired to C++ streams), and another method, say `juxtapose`, which is the one implemented in the suggestion, taking instead the `|` alias.

@alecandido I implemented this suggestion in 387a696ebb47793e0ef62c683a1c95d24393f63e and fixed the tests. If you agree with the implementation we can merge.

Reply:
I agree with your implementation.

The remaining part of the discussion has been moved to #1042